internal/backends/java/java.go

```diff
@@ -0,0 +1,201 @@
+	Version    string   `xml:"version"`
+}
+
+type Project struct {
+	XMLName      xml.Name     `xml:"project"`
+	ModelVersion string       `xml:"modelVersion"`
+	GroupId      string       `xml:"groupId"`
+	ArtifactId   string       `xml:"artifactId"`
+	Version      string       `xml:"version"`
+	Dependencies []Dependency `xml:"dependencies>dependency"`
+}
+
+var emptyProject = Project{
+	ModelVersion: "4.0.0",
+	GroupId:      "co.repl",
+	ArtifactId:   "artifact",
+	Version:      "0.0-SNAPSHOT",
+	Dependencies: []Dependency{},
+}
+
+var pkgNameRegexp = regexp.MustCompile("^([^:]+):([^:]+)") // groupid:artifactid
+
+// javaPatterns is the FilenamePatterns value for JavaBackend.
+var javaPatterns = []string{"*.java"}
+
+func readProjectOrMakeEmpty(path string) Project {
+	var project Project
+	if util.Exists("pom.xml") {
+		xmlbytes, err := ioutil.ReadFile("pom.xml")
+		if err != nil {
+			util.Die("error reading pom.xml: %s", err)
+		}
+		err = xml.Unmarshal(xmlbytes, &project)
+		if err != nil {
+			util.Die("error unmarshalling pom.xml: %s", err)
+		}
+	} else {
+		project = emptyProject
+	}
+	return project
+}
+
+const pomdotxml = "pom.xml"
+
+// JavaBackend is the UPM language backend for Java using Maven.
+var JavaBackend = api.LanguageBackend{
+	Name:             "java-maven",
+	Specfile:         pomdotxml,
+	Lockfile:         pomdotxml,
+	FilenamePatterns: javaPatterns,
+	Quirks:           api.QuirksAddRemoveAlsoLocks,
+	GetPackageDir: func() string {
+		return "target/dependency"
+	},
+	Search: func(query string) []api.PkgInfo {
+		searchDocs, err := Search(query)
+		if err != nil {
+			util.Log("error searching maven %s", err)
+			return []api.PkgInfo{}
+		}
+		pkgInfos := []api.PkgInfo{}
+		for _, searchDoc := range searchDocs {
+			pkgInfo := api.PkgInfo{
+				Name:    fmt.Sprintf("%s:%s", searchDoc.Group, searchDoc.Artifact),
+				Version: searchDoc.Version,
+			}
+			pkgInfos = append(pkgInfos, pkgInfo)
+		}
+		return pkgInfos
+	},
+	Info: func(name api.PkgName) api.PkgInfo {
+		return api.PkgInfo{
+			Name: string(name),
+		}
+	},
+	Add: func(pkgs map[api.PkgName]api.PkgSpec) {
+		project := readProjectOrMakeEmpty(pomdotxml)
+		existingDependencies := map[api.PkgName]api.PkgVersion{}
+		for _, dependency := range project.Dependencies {
+			pkgName := api.PkgName(strings.Join([]string{dependency.GroupId, dependency.ArtifactId}, ":"))
+			pkgVersion := api.PkgVersion(dependency.Version)
+			existingDependencies[pkgName] = pkgVersion
+		}
+
+		newDependencies := []Dependency{}
+		for pkgName, pkgSpec := range pkgs {
+			submatches := pkgNameRegexp.FindStringSubmatch(string(pkgName))
+			if nil == submatches {
+				util.Die("package name %s does not match groupid:artifactid pattern")
+			} else {
+				groupId := submatches[1]
+				artifactId := submatches[2]
+				if _, ok := existingDependencies[pkgName]; ok {
+					// this package is already in the lock file
+				} else {
+					var versionString string
```

Comment:
things get really nested in here, try to avoid deep indentation. We **generally** try to stay <= 80 columns with tabs being 4 spaces.

Reply:
I pulled out functions longer than a line and broke up some looong lines.